### PR TITLE
Properly parse address member.

### PR DESCRIPTION
**Pull request by chriseth**
References https://github.com/ethereum/solidity/issues/10905

### CODE REVIEW

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -961,6 +961,14 @@ ASTPointer<Identifier> Parser::parseIdentifier()
 	return nodeFactory.createNode<Identifier>(expectIdentifierToken());
 }
 
+ASTPointer<Identifier> Parser::parseIdentifierIncludingAddress()
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp]: Why not `IdentifierOrAddress` given that is what's used the scanner/ast? If the parser uses `including` more frequently, then of course that is a better choice.

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -2204,6 +2206,23 @@ ASTPointer<ASTString> Parser::expectIdentifierToken()
...
+{
+	ASTPointer<ASTString> result;
+	if (m_scanner->currentToken() == Token::Address)
+	{
+		result = make_shared<ASTString>("address");
+		m_scanner->next();
+	}
+	else
+	{
+		// do not advance on success
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp]: This comment is confusing, given `getLiteralAndAdvance` does it already? I think the comment is warning that no `next()` is required after `getLiteralAndAdvance`, but I would just get rid of the comment.

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: The comment explains the false flag. I will move it.

**Comment by chriseth** [comment]
Updated.

File: libsolidity/parsing/Parser.cpp [hunk 3]
```diff
@@ -2199,11 +2201,26 @@ ASTPointer<ParameterList> Parser::createEmptyParameterList()
...
-	// do not advance on success
-	expectToken(Token::Identifier, false);
+	expectToken(Token::Identifier, false /* do not advance */);
 	return getLiteralAndAdvance();
 }
 
+ASTPointer<ASTString> Parser::expectIdentifierTokenOrAddress()
+{
+	ASTPointer<ASTString> result;
+	if (m_scanner->currentToken() == Token::Address)
```
Comment (axic) [inline libsolidity/parsing/Parser.cpp:2211]: Do we actually have any other such tokens which are valid as an identifier? Perhaps we'll need to extend this. Okay to merge this PR as is to limit the scope, but probably worth looking into this.

File: test/libsolidity/semanticTests/functionTypes/address_member.sol [hunk 4]
```diff
@@ -0,0 +1,12 @@
+contract C {
+    function f() public view returns (address a1, address a2) {
+        a1 = this.f.address;
+        this.f.address;
+        [this.f.address][0];
```
Comment (axic) [inline test/libsolidity/semanticTests/functionTypes/address_member.sol:5]: What is this line doing? :) Creating a compile-time array with a single member and indexing it?

**Review by axic — APPROVED** [review]

File: test/libsolidity/semanticTests/functionTypes/address_member.sol (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/functionTypes/address_member.sol:5]: yes

File: libsolidity/parsing/Parser.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:2211]: Yes, we can extend it, but currently, nothing comes to my mind.

**Pull request merged**